// react-native-ui-kitten/ui/bottomNavigation/bottomNavigationTab.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { StyleSheet, TouchableOpacity, } from 'react-native';
import { styled, } from '../../theme';
import { Text, } from '../text/text.component';
import { isValidString } from '../support/services';
/**
 * `BottomNavigationTab` component is a part of the `BottomNavigation` component.
 * `BottomNavigation` tabs should be wrapped in BottomNavigation to provide usable component.
 * See usage examples at `BottomNavigation` component documentation.
 *
 * @extends React.Component
 *
 * @property {boolean} selected - Determines whether component is selected.
 *
 * @property {string} title - Determines the title of the tab.
 *
 * @property {StyleProp<TextStyle>} titleStyle - Customizes title style.
 *
 * @property {(style: StyleType) => React.ReactElement<ImageProps>} icon - Determines the icon of the tab.
 *
 * @property {(selected: boolean) => void} onSelect - Triggered on select value.
 *
 * @property TouchableOpacityProps
 *
 * @property StyledComponentProps
 *
 * @overview-example Simple Usage
 *
 * ```
 * import React from 'react';
 * import { BottomNavigationTab } from '../..';
 *
 * export const BottomNavigationTabShowcase = (props) => (
 *   <BottomNavigationTab title='Dashboard' />
 * );
 * ```
 *
 * @overview-example With Icon
 *
 * ```
 * // IMPORTANT: To use Icon component make sure to follow this guide:
 * // https://akveo.github.io/react-native-ui-kitten/docs/guides/eva-icons
 *
 * import React from 'react';
 * import { BottomNavigationTab, Icon } from '../..';
 *
 * const DashboardIcon = (style) => (
 *   <Icon {...style} name='layout' />
 * );
 *
 * export const BottomNavigationTabShowcase = (props) => (
 *   <BottomNavigationTab title='Dashboard' icon={DashboardIcon} />
 * );
 * ```
 *
 * @example Using Asset Icons
 *
 * ```
 * import React from 'react';
 * import { Image } from 'react-native';
 * import { BottomNavigationTab } from '../..';
 *
 * const DashboardIcon = (style) => (
 *   <Image style={style} source={require('path-to-assets/local-image.png')} />
 * );
 *
 * export const BottomNavigationTabShowcase = (props) => (
 *   <BottomNavigationTab title='Dashboard' icon={DashboardIcon} />
 * );
 * ```
 *
 * @example Inline Styling
 *
 * ```
 * import React from 'react';
 * import { StyleSheet } from 'react-native';
 * import { BottomNavigationTab } from '../..';
 *
 * export const BottomNavigationTabShowcase = (props) => (
 *   <BottomNavigationTab
 *     style={styles.tab}
 *     titleStyle={styles.tabTitle}
 *     title='Dashboard'
 *   />
 * );
 *
 * const styles = StyleSheet.create({
 *   tab: { backgroundColor: 'white' },
 *   tabTitle: { color: 'black' },
 * });
 * ```
 */
export class BottomNavigationTabComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.onPress = () => {
            if (this.props.onSelect) {
                this.props.onSelect(!this.props.selected);
            }
        };
        this.getComponentStyle = (source) => {
            const { iconWidth, iconHeight, iconMarginVertical, iconTintColor, textMarginVertical, textFontFamily, textFontSize, textLineHeight, textFontWeight, textColor } = source, containerParameters = __rest(source, ["iconWidth", "iconHeight", "iconMarginVertical", "iconTintColor", "textMarginVertical", "textFontFamily", "textFontSize", "textLineHeight", "textFontWeight", "textColor"]);
            return {
                container: containerParameters,
                text: {
                    marginVertical: textMarginVertical,
                    fontFamily: textFontFamily,
                    fontSize: textFontSize,
                    lineHeight: textLineHeight,
                    fontWeight: textFontWeight,
                    color: textColor,
                },
                icon: {
                    width: iconWidth,
                    height: iconHeight,
                    marginVertical: iconMarginVertical,
                    tintColor: iconTintColor,
                },
            };
        };
        this.renderIconElement = (style) => {
            const iconElement = this.props.icon(style);
            return React.cloneElement(iconElement, {
                key: 1,
                style: [style, styles.icon, iconElement.props.style],
            });
        };
        this.renderTitleElement = (style) => {
            const { title, titleStyle } = this.props;
            return (<Text key={2} style={[style, styles.text, titleStyle]}>
        {title}
      </Text>);
        };
        this.renderComponentChildren = (style) => {
            const { icon, title } = this.props;
            return [
                icon && this.renderIconElement(style.icon),
                isValidString(title) && this.renderTitleElement(style.text),
            ];
        };
    }
    render() {
        const _a = this.props, { style, themedStyle } = _a, restProps = __rest(_a, ["style", "themedStyle"]);
        const _b = this.getComponentStyle(themedStyle), { container } = _b, componentStyles = __rest(_b, ["container"]);
        const [iconElement, titleElement] = this.renderComponentChildren(componentStyles);
        return (<TouchableOpacity {...restProps} style={[container, styles.container, style]} activeOpacity={1.0} onPress={this.onPress}>
        {iconElement}
        {titleElement}
      </TouchableOpacity>);
    }
}
BottomNavigationTabComponent.styledComponentName = 'BottomNavigationTab';
const styles = StyleSheet.create({
    container: {
        justifyContent: 'center',
        alignItems: 'center',
    },
    text: {},
    icon: {},
});
export const BottomNavigationTab = styled(BottomNavigationTabComponent);
