// react-native-ui-kitten/ui/calendar/components/picker/calendarPickerCell.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { StyleSheet, TouchableOpacity, } from 'react-native';
import { styled, } from '../../../../theme';
class CalendarPickerCellComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.onPress = () => {
            if (this.props.onSelect) {
                this.props.onSelect(this.props.date);
            }
        };
        this.getContainerBorderRadius = (borderRadius) => {
            const { firstRangeItem, lastRangeItem } = this.props;
            if (firstRangeItem) {
                return {
                    borderBottomLeftRadius: borderRadius,
                    borderBottomRightRadius: 0,
                    borderTopLeftRadius: borderRadius,
                    borderTopRightRadius: 0,
                };
            }
            if (lastRangeItem) {
                return {
                    borderBottomLeftRadius: 0,
                    borderBottomRightRadius: borderRadius,
                    borderTopLeftRadius: 0,
                    borderTopRightRadius: borderRadius,
                };
            }
            return {};
        };
        this.getComponentStyle = (source) => {
            const { contentBorderWidth, contentBorderRadius, contentBorderColor, contentBackgroundColor, contentTextFontSize, contentTextLineHeight, contentTextFontWeight, contentTextColor, contentTextFontFamily, borderRadius } = source, containerParameters = __rest(source, ["contentBorderWidth", "contentBorderRadius", "contentBorderColor", "contentBackgroundColor", "contentTextFontSize", "contentTextLineHeight", "contentTextFontWeight", "contentTextColor", "contentTextFontFamily", "borderRadius"]);
            return {
                container: Object.assign(Object.assign({}, containerParameters), this.getContainerBorderRadius(borderRadius)),
                contentContainer: {
                    borderWidth: contentBorderWidth,
                    borderRadius: contentBorderRadius,
                    borderColor: contentBorderColor,
                    backgroundColor: contentBackgroundColor,
                },
                contentText: {
                    fontSize: contentTextFontSize,
                    fontWeight: contentTextFontWeight,
                    lineHeight: contentTextLineHeight,
                    color: contentTextColor,
                    fontFamily: contentTextFontFamily,
                },
            };
        };
        this.renderContentElement = (source, style) => {
            return source && source(this.props.date, {
                container: style.contentContainer,
                text: style.contentText,
            });
        };
    }
    shouldComponentUpdate(nextProps) {
        if (nextProps.shouldComponentUpdate) {
            return nextProps.shouldComponentUpdate(this.props, nextProps);
        }
        return true;
    }
    render() {
        const _a = this.props, { style, themedStyle, date, bounding, children } = _a, restProps = __rest(_a, ["style", "themedStyle", "date", "bounding", "children"]);
        const _b = this.getComponentStyle(themedStyle), { container } = _b, componentStyles = __rest(_b, ["container"]);
        return (<TouchableOpacity activeOpacity={1.0} onPress={this.onPress} {...restProps} style={[container, styles.container, style]}>
        {this.renderContentElement(children, componentStyles)}
      </TouchableOpacity>);
    }
}
CalendarPickerCellComponent.styledComponentName = 'CalendarCell';
CalendarPickerCellComponent.defaultProps = {
    selected: false,
    today: false,
};
const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
});
export const CalendarPickerCell = styled(CalendarPickerCellComponent);
