// react-native-ui-kitten/ui/select/selectGroupOption.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { View, } from 'react-native';
import { styled, } from '../../theme';
import { SelectOption, } from './selectOption.component';
class SelectGroupOptionComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.getComponentStyle = (source) => {
            const { itemPaddingHorizontal } = source, containerStyles = __rest(source, ["itemPaddingHorizontal"]);
            return {
                container: containerStyles,
                item: {
                    paddingHorizontal: itemPaddingHorizontal,
                },
            };
        };
        this.getMainItemStatus = (subItemsSelectedStatusArray) => {
            const someSelected = subItemsSelectedStatusArray
                .some((item) => item === true);
            const everySelected = subItemsSelectedStatusArray
                .every((item) => item === true);
            switch (true) {
                case (someSelected && !everySelected):
                    return { selected: true, indeterminate: true };
                case !someSelected:
                    return { selected: false, indeterminate: false };
                case everySelected:
                    return { selected: true, indeterminate: false };
            }
        };
        this.renderSubItem = (option, index) => {
            const _a = this.props, { item, renderItem, strategy } = _a, restProps = __rest(_a, ["item", "renderItem", "strategy"]);
            const returningOption = {
                item: option,
                index: index,
                separators: null,
            };
            const selected = strategy.isSelected(option);
            return renderItem ? renderItem(returningOption) : (<SelectOption {...restProps} selected={selected} item={option}/>);
        };
        this.renderSubItemsElements = () => {
            const { item, themedStyle } = this.props;
            const { item: itemStyle } = this.getComponentStyle(themedStyle);
            return item.items
                .map((option, index) => {
                const element = this.renderSubItem(option, index);
                return React.cloneElement(element, Object.assign(Object.assign({}, option), { style: [element.props.style, itemStyle], key: index }));
            });
        };
        this.renderMultiSelectMainElement = (subItemsElements) => {
            const _a = this.props, { item } = _a, restProps = __rest(_a, ["item"]);
            const subItemsSelectedStatusArray = subItemsElements
                .map((subItem) => subItem.props.selected);
            const itemStatus = this.getMainItemStatus(subItemsSelectedStatusArray);
            return (<SelectOption {...restProps} {...itemStatus} item={item}/>);
        };
        this.renderDefaultMainElement = () => {
            const { item } = this.props;
            return (<SelectOption disabled={true} item={item}/>);
        };
        this.renderMainElement = (subItemsElements) => {
            const { multiSelect } = this.props;
            return multiSelect ? this.renderMultiSelectMainElement(subItemsElements) : this.renderDefaultMainElement();
        };
    }
    render() {
        const { themedStyle } = this.props;
        const { container } = this.getComponentStyle(themedStyle);
        const subItemsElements = this.renderSubItemsElements();
        const mainElement = this.renderMainElement(subItemsElements);
        return (<View style={container}>
        {mainElement}
        {subItemsElements}
      </View>);
    }
}
SelectGroupOptionComponent.styledComponentName = 'SelectGroupOption';
export const SelectGroupOption = styled(SelectGroupOptionComponent);
