// react-native-ui-kitten/ui/select/selection.strategy.js
export class MultiSelectStrategy {
    constructor(options) {
        if (Array.isArray(options)) {
            this.selectedOption = options;
        }
    }
    select(option, callback) {
        const subOptionsExist = this.areThereSubOptions(option);
        if (subOptionsExist) {
            this.selectOptionWithSubOptions(option);
        }
        else {
            this.selectDefaultOption(option);
        }
        return this.selectedOption;
    }
    selectDefaultOption(option) {
        const optionAlreadyExist = this.selectedOption
            .some((item) => {
            return item === option;
        });
        if (optionAlreadyExist) {
            this.removeOption(option);
        }
        else {
            this.selectedOption.push(option);
        }
    }
    selectOptionWithSubOptions(option) {
        const subOptionsAlreadyExist = this.selectedOption
            .some((item) => {
            return option.items
                .some((subItem) => {
                return subItem === item;
            });
        });
        if (subOptionsAlreadyExist) {
            option.items.forEach((subItem) => this.removeOption(subItem));
        }
        else {
            const enabledItems = option.items
                .filter((item) => {
                return !item.disabled;
            });
            this.selectedOption = this.selectedOption.concat(enabledItems);
        }
    }
    getPlaceholder(placeholder) {
        if (this.isSelectedOptionExist()) {
            return this.selectedOption
                .map((item) => {
                return item && item.text;
            })
                .join(', ');
        }
        else {
            return placeholder;
        }
    }
    isSelected(item) {
        return this.selectedOption
            .some((option) => {
            return option === item;
        });
    }
    isSelectedOptionExist() {
        return this.selectedOption && this.selectedOption.length !== 0;
    }
    removeOption(option) {
        const index = this.selectedOption
            .findIndex((item) => {
            return item === option;
        });
        if (index !== -1) {
            this.selectedOption.splice(index, 1);
        }
    }
    areThereSubOptions(option) {
        return option.items && option.items.length !== 0;
    }
}
export class SingleSelectStrategy {
    constructor(options) {
        if (!Array.isArray(options)) {
            this.selectedOption = options;
        }
    }
    select(option, callback) {
        this.selectedOption = option;
        if (callback) {
            callback();
        }
        return this.selectedOption;
    }
    getPlaceholder(placeholder) {
        if (this.selectedOption) {
            return this.selectedOption.text;
        }
        else {
            return placeholder;
        }
    }
    isSelected(item) {
        if (this.hasOptionSubItems(item)) {
            return item.items.some((option) => {
                return this.isSelected(option);
            });
        }
        else {
            return this.selectedOption === item;
        }
    }
    hasOptionSubItems(option) {
        return option.items && option.items.length !== 0;
    }
}
