// react-native-ui-kitten/ui/select/selectOption.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import { StyleSheet, TouchableOpacity, View, } from 'react-native';
import { Interaction, styled, } from '../../theme';
import { Text, } from '../text/text.component';
import { CheckBox } from '../checkbox/checkbox.component';
class SelectOptionComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.onPress = (event) => {
            const { item, onPress } = this.props;
            this.props.dispatch([]);
            onPress(item, event);
        };
        this.onMultiSelectItemPress = (value) => {
            this.onPress(null);
        };
        this.onPressIn = (event) => {
            this.props.dispatch([Interaction.ACTIVE]);
            if (this.props.onPressIn) {
                this.props.onPressIn(this.props.item, event);
            }
        };
        this.onPressOut = (event) => {
            this.props.dispatch([]);
            if (this.props.onPressOut) {
                this.props.onPressOut(this.props.item, event);
            }
        };
        this.onLongPress = (event) => {
            if (this.props.onLongPress) {
                this.props.onLongPress(this.props.item, event);
            }
        };
        this.getComponentStyle = (source) => {
            const { textColor, textFontFamily, textFontSize, textFontWeight, textLineHeight, textMarginHorizontal, multiSelectBackgroundColor, multiSelectTextColor } = source, containerStyles = __rest(source, ["textColor", "textFontFamily", "textFontSize", "textFontWeight", "textLineHeight", "textMarginHorizontal", "multiSelectBackgroundColor", "multiSelectTextColor"]);
            return {
                container: containerStyles,
                multiSelectContainer: Object.assign(Object.assign({}, containerStyles), { backgroundColor: multiSelectBackgroundColor }),
                text: {
                    color: textColor,
                    fontFamily: textFontFamily,
                    fontSize: textFontSize,
                    fontWeight: textFontWeight,
                    lineHeight: textLineHeight,
                    marginHorizontal: textMarginHorizontal,
                },
                multiSelectText: {
                    color: multiSelectTextColor,
                    fontFamily: textFontFamily,
                    fontSize: textFontSize,
                    fontWeight: textFontWeight,
                    lineHeight: textLineHeight,
                    marginHorizontal: textMarginHorizontal,
                },
            };
        };
        this.renderTextElement = (style) => {
            const { item } = this.props;
            return (<Text style={[style, styles.text, item.textStyle]}>
        {item.text}
      </Text>);
        };
        this.renderDefaultItem = () => {
            const _a = this.props, { themedStyle, style, item } = _a, restProps = __rest(_a, ["themedStyle", "style", "item"]);
            const { container, text } = this.getComponentStyle(themedStyle);
            const textElement = this.renderTextElement(text);
            return (<TouchableOpacity {...restProps} activeOpacity={1.0} style={[styles.container, container, style]} onPress={this.onPress} onPressIn={this.onPressIn} onPressOut={this.onPressOut} onLongPress={this.onLongPress}>
        {textElement}
      </TouchableOpacity>);
        };
        this.renderMultiSelectItem = () => {
            const _a = this.props, { disabled, item, themedStyle, selected, style, indeterminate } = _a, restProps = __rest(_a, ["disabled", "item", "themedStyle", "selected", "style", "indeterminate"]);
            const { multiSelectContainer, multiSelectText } = this.getComponentStyle(themedStyle);
            return (<View {...restProps} style={[styles.container, multiSelectContainer, style]}>
        <CheckBox text={item.text} textStyle={[multiSelectText, item.textStyle]} disabled={disabled} checked={selected} indeterminate={indeterminate} onChange={this.onMultiSelectItemPress}/>
      </View>);
        };
    }
    render() {
        const { multiSelect } = this.props;
        return multiSelect ? this.renderMultiSelectItem() : this.renderDefaultItem();
    }
}
SelectOptionComponent.styledComponentName = 'SelectOption';
const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    icon: {},
    text: {},
});
export const SelectOption = styled(SelectOptionComponent);
